src/node_code_integrity.cc

```diff
@@ -0,0 +1,279 @@
+#include "node_code_integrity.h"
+#include "v8.h"
+#include "node.h"
+#include "env-inl.h"
+#include "node_external_reference.h"
+
+namespace node {
+
+using v8::Boolean;
+using v8::Context;
+using v8::FunctionCallbackInfo;
+using v8::Local;
+using v8::Object;
+using v8::Value;
+
+namespace codeintegrity {
+
+#ifdef _WIN32
+static bool isWldpInitialized = false;
+static pfnWldpCanExecuteFile WldpCanExecuteFile;
+static pfnWldpGetApplicationSettingBoolean WldpGetApplicationSettingBoolean;
+static pfnWldpQuerySecurityPolicy WldpQuerySecurityPolicy;
+static PCWSTR NODEJS = L"Node.js";
+static PCWSTR ENFORCE_CODE_INTEGRITY_SETTING_NAME = L"EnforceCodeIntegrity";
+static PCWSTR DISABLE_INTERPRETIVE_MODE_SETTING_NAME =
+  L"DisableInteractiveMode";
+
+void InitWldp(Environment* env) {
+  if (isWldpInitialized) {
+    return;
+  }
+
+  HMODULE wldp_module = LoadLibraryExA(
+      "wldp.dll",
+      nullptr,
+      LOAD_LIBRARY_SEARCH_SYSTEM32);
+
+  if (wldp_module == nullptr) {
+    return env->ThrowError("Unable to load wldp.dll");
+  }
+
+  WldpCanExecuteFile =
+    (pfnWldpCanExecuteFile)GetProcAddress(
+      wldp_module,
+      "WldpCanExecuteFile");
+
+  WldpGetApplicationSettingBoolean =
+    (pfnWldpGetApplicationSettingBoolean)GetProcAddress(
+      wldp_module,
+      "WldpGetApplicationSettingBoolean");
+
+  WldpQuerySecurityPolicy =
+    (pfnWldpQuerySecurityPolicy)GetProcAddress(
+      wldp_module,
+      "WldpQuerySecurityPolicy");
+
+  isWldpInitialized = true;
+}
+
+static void IsFileTrustedBySystemCodeIntegrityPolicy(
+  const FunctionCallbackInfo<Value>& args) {
+  CHECK_EQ(args.Length(), 1);
+  CHECK(args[0]->IsString());
+
+  Environment* env = Environment::GetCurrent(args);
+  if (!isWldpInitialized) {
+    InitWldp(env);
+  }
+
+  BufferValue path(env->isolate(), args[0]);
+  if (*path == nullptr) {
+    return env->ThrowError("path cannot be empty");
+  }
+
+  HANDLE hFile = CreateFileA(
+    *path,
+    GENERIC_READ,
+    FILE_SHARE_READ,
+    nullptr,
+    OPEN_EXISTING,
+    FILE_ATTRIBUTE_NORMAL,
+    nullptr);
+
+  if (hFile == INVALID_HANDLE_VALUE || hFile == nullptr) {
+    return env->ThrowError("Unable to open file");
+  }
+
+  const GUID wldp_host_other = WLDP_HOST_OTHER;
+  WLDP_EXECUTION_POLICY result;
+  HRESULT hr = WldpCanExecuteFile(
```

Comment:
Are those calls synchronous right? It seems to block the event loop on every check making the system useless from a nodejs perspective

Reply:
Yes. These calls are synchronous, and there are no async versions. Is there a way to address this without the platform adding an async version?

This call will only be made on systems that have opted into code integrity enforcement. I can reach out to the Windows platform team and see if there's opportunity to add an async version if this is an issue.

At the moment, is it acceptable to say there will be performance hits when using this feature?

Reply:
I think it will make this feature hard to use considering Node.js constraints, we can measure but I suspect it will make things quite slow